oshmem/mca/spml/ucx/spml_ucx.c

```diff
@@ -126,6 +126,140 @@ static ucp_request_param_t mca_spml_ucx_request_param_b = {
 };
 #endif
 
+unsigned
+mca_spml_ucx_mem_map_flags_symmetric_rkey(struct mca_spml_ucx *spml_ucx)
+{
+#if HAVE_DECL_UCP_MEM_MAP_SYMMETRIC_RKEY
+    if (spml_ucx->symmetric_rkey) {
+        return UCP_MEM_MAP_SYMMETRIC_RKEY;
+    }
+#endif
+
+    return 0;
+}
+
+int mca_spml_ucx_rkey_store_create(mca_spml_ucx_rkey_store_t *store,
+                                   ucp_worker_h worker,
+                                   int size)
+{
+    store->array = calloc(size, sizeof(*store->array));
+    if (store->array == NULL) {
+        return OSHMEM_ERR_OUT_OF_RESOURCE;
+    }
+
+    store->size   = size;
+    store->count  = 0;
+    store->worker = worker;
+    return OSHMEM_SUCCESS;
+}
+
+void mca_spml_ucx_rkey_store_destroy(mca_spml_ucx_rkey_store_t *store)
+{
+    int i;
+
+    for (i = 0; i < store->count; i++) {
+        assert(store->array[i].refcnt == 0);
+        ucp_rkey_destroy(store->array[i].rkey);
+    }
+
+    free(store->array);
+    store->array = NULL;
+    store->count = 0;
+    store->size  = 0;
```

Comment:
It can be done with `opal_value_array_t`. But we need to insert values at random indexes to keep array ordered, and for that we need to add an opal_value_array insert function (at specific indexes with `memmove()`) in `opal/class/opal_value_array.h`.

Since none of the existing potential users in oshmem use `opal_*_array_t`, are we sure to want to go that route?

Reply:
I see, so better keep as is